prov/rxd/src/rxd_av.c

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright (c) 2015-2016 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "rxd.h"
+
+fi_addr_t rxd_av_get_dg_addr(struct rxd_av *av, fi_addr_t fi_addr)
+{
+	uint64_t *dg_idx;
+	dg_idx = ofi_av_get_addr(&av->util_av, (int) fi_addr);
+	return *dg_idx;
+}
+
+fi_addr_t rxd_av_get_fi_addr(struct rxd_av *av, fi_addr_t dg_addr)
+{
+	return (fi_addr_t) ofi_av_lookup_index(&av->util_av,
+					      &dg_addr, (int) dg_addr);
+}
+
+int rxd_av_insert_dg_av(struct rxd_av *av, const void *addr)
+{
+	int ret;
+	fastlock_acquire(&av->lock);
+	ret = fi_av_insert(av->dg_av, addr, 1, NULL, 0, NULL);
+	if (ret != 1)
+		goto out;
+	av->dg_av_used++;
+out:
+	fastlock_release(&av->lock);
+	return ret;
+}
+
+int rxd_dg_av_reverse_lookup(struct rxd_av *av, uint64_t start_idx,
+			      const void *addr, size_t addrlen, uint64_t *idx)
+{
+	int ret;
+	size_t i, len;
+	void *curr_addr;
+
+	len = addrlen;
+	curr_addr = calloc(1, av->addrlen);
+	if (!curr_addr)
+		return -FI_ENOMEM;
+
+	for (i = 0; i < av->dg_av_used; i++) {
+		ret = fi_av_lookup(av->dg_av, (i + start_idx) % av->dg_av_used,
+				   curr_addr, &len);
+		if (ret)
+			continue;
+		if (memcmp(curr_addr, addr, len) == 0) {
```

Comment:
The comparison is only valid when len == addrlen, otherwise the result can be true while it shouldn't.